lib/rust/ensogl/core/src/display/scene.rs

```diff
@@ -289,19 +293,35 @@ impl Mouse {
         if new_target != current_target {
             self.target.set(new_target);
             if let Some(event) = (*self.last_move_event.borrow()).clone() {
-                self.pointer_target_registry.with_mouse_target(current_target, |t, d| {
-                    t.mouse_out.emit(());
+                let new_hovered = self
+                    .pointer_target_registry
+                    .with_mouse_target(new_target, |_, d| d.rev_parent_chain())
+                    .unwrap_or_default();
+                let currently_hovered_weak = self.hovered_objects.take().into_iter();
+                let currently_hovered =
+                    currently_hovered_weak.filter_map(|w| w.upgrade()).collect_vec();
+                let left = currently_hovered.iter().filter(|t| !new_hovered.contains(t));
+                for d in left {
                     let out_event = event.clone().unchecked_convert_to::<mouse::Out>();
                     let leave_event = event.clone().unchecked_convert_to::<mouse::Leave>();
                     d.emit_event(out_event);
                     d.emit_event_without_bubbling(leave_event);
-                });
-                self.pointer_target_registry.with_mouse_target(new_target, |t, d| {
-                    t.mouse_over.emit(());
+                }
+                let entered = new_hovered.iter().filter(|t| !currently_hovered.contains(t));
```

Comment:
The `O(N x M)` scaling of these `Vec` set operations worries me a little. The display object hierarchy is already deep, and it could reach a depth where it suddenly becomes a performance problem. `HashSet`s would be safer.

Reply:
I just assumed that even if object hierarchy is 50 or so levels deep, iterating over 50 elements may be still better than allocating additional `HashMaps`.

But actually, we can easily avoid allocation while improving complexity by just sorting the vecs. I'll push changes soon.